File: src/main/output/prettify.ts
import { error } from './theme';

export function prettifyAsync(
  fn: (js: string) => Promise<void>,
): (js: string) => Promise<void> {
  return async (js: string) => {
    try {
      await fn(js);
    } catch (e: unknown) {
      console.error(error('Fatal:'));
      if (e instanceof Error) {
        console.error(error(e.message));
      } else {
        console.error(error(String(e)));
      }
    }
  };
}

File: src/main/kotlin/output/Prettify.kt
package com.climat.output

import process

fun <T, U> prettify(fn: (arg1: T, arg2: U) -> Unit): (arg1: T, arg2: U) -> Unit =
    { a1, a2 ->
        try {
            fn(a1, a2)
        } catch (ex: Throwable) {
            ex.handle()
        }
    }

fun prettify(fn: () -> Unit): (() -> Unit) =
    {
        try {
            fn()
        } catch (ex: Throwable) {
            ex.handle()
        }
    }

fun Throwable.handle() {
    console.error(error("Fatal:"))
    console.error(error(message ?: stackTraceToString()))
    console.error(error(stackTraceToString()))
    process.exit(-1)
}
